Sherlock/test.cc

```diff
@@ -507,3 +512,81 @@ TEST(Sherlock, ParsesFromFile) {
       CompareValuesMixedWithTerminate(d.results_, expected_values, SherlockTestProcessor::kTerminateStr))
       << d.results_;
 }
+
+TEST(Sherlock, SubscribeWithFilterByType) {
+  using namespace sherlock_unittest;
+
+  struct CollectorImpl {
+    CollectorImpl() = delete;
+    CollectorImpl(const CollectorImpl&) = delete;
+    CollectorImpl(CollectorImpl&&) = delete;
+
+    explicit CollectorImpl(size_t expected_count) : expected_count_(expected_count) {}
+
+    EntryResponse operator()(const Variant<Record, AnotherRecord>& entry, idxts_t, idxts_t) {
+      results_.push_back(JSON<JSONFormat::Minimalistic>(entry));
+      return results_.size() == expected_count_ ? EntryResponse::Done : EntryResponse::More;
+    }
+
+    EntryResponse operator()(const Record& record, idxts_t, idxts_t) {
+      results_.push_back("X=" + current::ToString(record.x));
+      return results_.size() == expected_count_ ? EntryResponse::Done : EntryResponse::More;
+    }
+
+    EntryResponse operator()(const AnotherRecord& another_record, idxts_t, idxts_t) {
+      results_.push_back("Y=" + current::ToString(another_record.y));
+      return results_.size() == expected_count_ ? EntryResponse::Done : EntryResponse::More;
+    }
+
+    TerminationResponse Terminate() const { return TerminationResponse::Wait; }
+
+    std::vector<std::string> results_;
+    const size_t expected_count_;
+  };
+
+  auto stream = current::sherlock::Stream<Variant<Record, AnotherRecord>>();
+  for (int i = 1; i <= 5; ++i) {
+    current::time::SetNow(std::chrono::microseconds(i));
+    if (i & 1) {
+      stream.Publish(Record(i));
+    } else {
+      stream.Publish(AnotherRecord(i));
+    }
+  }
+
+  {
+    using Collector = current::ss::StreamSubscriber<CollectorImpl, Variant<Record, AnotherRecord>>;
+    static_assert(current::ss::IsStreamSubscriber<Collector, Variant<Record, AnotherRecord>>::value, "");
+    static_assert(!current::ss::IsStreamSubscriber<Collector, Record>::value, "");
+    static_assert(!current::ss::IsStreamSubscriber<Collector, AnotherRecord>::value, "");
+
+    Collector c(5);
+    stream.Subscribe(c).Join();
+    EXPECT_EQ(
+        "{\"Record\":{\"x\":1}} {\"AnotherRecord\":{\"y\":2}} {\"Record\":{\"x\":3}} "
+        "{\"AnotherRecord\":{\"y\":4}} {\"Record\":{\"x\":5}}",
+        Join(c.results_, ' '));
+  }
+
+  {
+    using Collector = current::ss::StreamSubscriber<CollectorImpl, Record>;
+    static_assert(!current::ss::IsStreamSubscriber<Collector, Variant<Record, AnotherRecord>>::value, "");
+    static_assert(current::ss::IsStreamSubscriber<Collector, Record>::value, "");
+    static_assert(!current::ss::IsStreamSubscriber<Collector, AnotherRecord>::value, "");
+
+    Collector c(3);
+    stream.Subscribe<Record>(c).Join();
```

Comment:
@mzhurovich Actually, the semantics can be even easier -- as `.Subscribe()` requires a `current::ss::StreamSubscriber`, the latter may just expose some `entry_t` of its own. Talk tomorrow / tonight :-)